### DISCUSSION TITLE
[11일차-지원] 웹사이트 성능 최적화를 위해서는 어떤 방법이 있을까요?

### DISCUSSION THREAD

**Discussion by smarfy99** [zzangCS/zzangInterview#38] [프론트 면접 질문]
웹사이트 성능 최적화를 위해서는 어떤 방법이 있을 까요?

**Comment by ez1n** [comment]
브라우저는 HTML파일을 읽어온 후 상단부터 차례대로 로드된다.
이 과정에서 style 태그나 script 태그를 만나면 페이지 렌더링이 일시적으로 차단된다. 이 때 일시적으로 화면이 나타나지 않는 현상이 나타나고 이러한 현상을 최소화하여 성능을 최적화할 수 있다.

성능 최적화는 렌더링과 로딩 과정을 종합적으로 최적화하는 것을 말한다.

1. 렌더링 최적화
- CSS 최적화
⇒ 리플로우가 일어나면 브라우저가 전체 픽셀을 다시 계산해야 하기 때문에 리페인트 속성을 사용하여 스타일을 작성한다.
- 리플로우 : 레이아웃의 넓이, 높이, 위치 등에 영향을 주는 css 속성을 변경할 경우 레이아웃부터 다시 그려지게 되는 현상
 - 리페인트 : 레이아웃에 영향을 주지 않는 속성을 변경하면 레이아웃을 건너뛰고 페인트 작업부터 다시 수행하는 현상
⇒ 사용하지 않는 CSS 제거
⇒ 간결한 스타일 작성
- HTML 최적화
⇒ 인라인 스타일을 작성하면 웹페이지가 그려지면서 레이아웃에 영향을 미치며 추가로 리플로우를 발생시키기 때문에 인라인 스타일을 사용하는 것은 지양한다.

2. 로딩 최적화
- 스타일 시트는 head 하단에 입력하면 순차적으로 페이지를 로딩하여 페이지 로딩 속도를 향상시킨다.
- 파일을 다운로드하여 실행하기 전까지 브라우저는 아무것도 렌더링하지 않기 때문에 스크립트는 body 하단에서 불러온다.
- 이미지 최적화
⇒ picture 태그를 사용하여 type 속성을 통해 사용자 환경에 맞는 이미지를 제공한다.
⇒ img의 로딩 속성을 활용하여 브라우저 화면에 지연/병렬 로딩이 가능하다.
